### PR TITLE
openjdk: mark major version 12 through 16 as EOL

### PR DESCRIPTION
**Pull request by mweinelt**
Mark EOL versions with `knownVulnerabilities`.

https://endoflife.date/java

Related: #170825

I don't consider #178025 fixed until these packages are removed.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/openjdk/meta.nix [hunk 1]
```diff
@@ -1,8 +1,11 @@
-lib: with lib; {
+lib: version: with lib; {
   homepage = "https://openjdk.java.net/";
   license = licenses.gpl2Only;
   description = "The open-source Java Development Kit";
   maintainers = with maintainers; [ edwtjo asbachb ];
   platforms = [ "i686-linux" "x86_64-linux" "aarch64-linux" "armv7l-linux" "armv6l-linux" ];
   mainProgram = "java";
+  knownVulnerabilities = optionals (builtins.elem (versions.major version) [ "12" "13" "14" "15" "16" ]) [
+    "This OpenJDK version has reached it's end of life."
```
Comment (dotlambda) [inline pkgs/development/compilers/openjdk/meta.nix:9]:
```suggestion
    "This OpenJDK version has reached its end of life."
```

**Comment by mweinelt** [comment]
```
error: value is a set while a string was expected

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/lib/versions.nix:20:31:

           19|   */
           20|   major = v: builtins.elemAt (splitVersion v) 0;
             |                               ^
           21|

       … while evaluating 'major'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/lib/versions.nix:20:11:

           19|   */
           20|   major = v: builtins.elemAt (splitVersion v) 0;
             |           ^
           21|

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/development/compilers/openjdk/meta.nix:8:52:

            7|   mainProgram = "java";
            8|   knownVulnerabilities = optionals (builtins.elem (versions.major version) [ "12" "13" "14" "15" "16" ]) [
             |                                                    ^
            9|     "This OpenJDK version has reached its end of life."

       … while evaluating 'optionals'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/lib/lists.nix:271:5:

          270|     # List to return if condition is true
          271|     elems: if cond then elems else [];
             |     ^
          272|

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/development/compilers/openjdk/meta.nix:8:26:

            7|   mainProgram = "java";
            8|   knownVulnerabilities = optionals (builtins.elem (versions.major version) [ "12" "13" "14" "15" "16" ]) [
             |                          ^
            9|     "This OpenJDK version has reached its end of life."

       … while evaluating the attribute 'meta.knownVulnerabilities'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/development/compilers/openjdk/meta.nix:8:3:

            7|   mainProgram = "java";
            8|   knownVulnerabilities = optionals (builtins.elem (versions.major version) [ "12" "13" "14" "15" "16" ]) [
             |   ^
            9|     "This OpenJDK version has reached its end of life."

       … while evaluating 'isMarkedInsecure'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:55:22:

           54|
           55|   isMarkedInsecure = attrs: (attrs.meta.knownVulnerabilities or []) != [];
             |                      ^
           56|

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:78:7:

           77|   hasAllowedInsecure = attrs:
           78|     !(isMarkedInsecure attrs) ||
             |       ^
           79|     allowInsecurePredicate attrs ||

       … while evaluating 'hasAllowedInsecure'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:77:24:

           76|
           77|   hasAllowedInsecure = attrs:
             |                        ^
           78|     !(isMarkedInsecure attrs) ||

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:287:15:

          286|       { valid = false; reason = "unsupported"; errormsg = "is not supported on ‘${hostPlatform.system}’"; }
          287|     else if !(hasAllowedInsecure attrs) then
             |               ^
          288|       { valid = false; reason = "insecure"; errormsg = "is marked as insecure"; }

       … while evaluating 'checkValidity'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:272:19:

          271|   # Along with a boolean flag for each reason
          272|   checkValidity = attrs:
             |                   ^
          273|     {

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:296:18:

          295|   assertValidity = { meta, attrs }: let
          296|       validity = checkValidity attrs;
             |                  ^
          297|     in validity // {

       … while evaluating 'assertValidity'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/check-meta.nix:295:20:

          294|
          295|   assertValidity = { meta, attrs }: let
             |                    ^
          296|       validity = checkValidity attrs;

       … from call site

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:338:14:

          337|
          338|   validity = checkMeta { inherit meta attrs; };
             |              ^
          339|

       … while evaluating the attribute 'installPhase' of the derivation 'scala-2.13.8'

       at /home/hexa/.cache/nixpkgs-review/pr-170827-1/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 170827` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by mweinelt** [comment]
In `versions.json` 1.17 uses `javaVersion` 16, that's why it get's "removed" in a sense.

**Comment by jyooru** [comment]
Versions below 1.17 use Java 8 - why aren't they removed?

When you say removed, is it just marked as insecure or has it actually been removed?

**Comment by mweinelt** [comment]
The build job gets removed, because it has been transitively marked insecure.

**Comment by jyooru** [comment]
Oh, that makes sense.

**Comment by wamserma** [comment]
Result of `nixpkgs-review pr 170827` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Pull request merged**